Very useful tool! Can we get it to work on Mac OS?

Thanks @riggerthegeek, I was able to get this working using the Fedora docker container on my Mac.  Below is the output I get when running `sudo make build` on a mac directly:

```
sh ./scripts/pibuilder.sh
tr: Illegal byte sequence
./scripts/pibuilder.sh: line 15: md5sum: command not found
Archive:  ./cache/os..zip
  inflating: ./cache/unzip/2018-03-13-raspbian-stretch.img
./scripts/pibuilder.sh: line 84: kpartx: command not found
make: *** [build] Error 127
```

First two commands work @riggerthegeek! The `make build` step fails due to a permissions error:

```bash
make docker-run CMD="sh ./scripts/pibuilder.sh"
touch settings.sh
mkdir -p ./cache
mkdir -p ./ssh-keys
docker run \
		-it \
		--privileged \
		--rm \
		-v "/Users/nhippenmeyer/src/pibuilder/settings.sh:/opt/pibuilder/settings.sh" \
		-v "/Users/nhippenmeyer/src/pibuilder/cache:/opt/pibuilder/cache" \
		-v "/Users/nhippenmeyer/src/pibuilder/ssh-keys:/ssh-keys" \
		-v "/Users/nhippenmeyer/src/pibuilder/scripts:/opt/pibuilder/scripts" \
		-u 0 \
		riggerthegeek/pibuilder \
		sh ./scripts/pibuilder.sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0   398    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0Warning: Failed to create the file
Warning: ./cache/os.76a0dae971f3a690a6422b6babbf2840.zip: Permission denied
  0  297M    0  2920    0     0   2920      0 29:37:35  0:00:01 29:37:34  9700
curl: (23) Failed writing body (0 != 2920)
make[1]: *** [docker-run] Error 23
make: *** [build] Error 2
```

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,0 +1,26 @@
+FROM fedora:27
+
+# Setup the user
+RUN groupadd --gid 1000 pibuilder \
+  && useradd --uid 1000 --gid pibuilder --shell /bin/bash --create-home pibuilder
+
+# Install NodeJS
+RUN curl --silent --location https://rpm.nodesource.com/setup_8.x | bash - \
+  && dnf update -y \
+  && dnf install -y nodejs \
+  && dnf install -y gcc-c++ make \
+  && node --version \
+  && npm --version
+
+WORKDIR /opt/pibuilder
+ADD . .
+
+ARG SSH_KEY_DIR="/ssh-keys"
+
+VOLUME ${SSH_KEY_DIR}
+
+ENV SSH_KEY_DIR=${SSH_KEY_DIR}
+
+# Install dependencies
+RUN dnf install -y git kpartx openssh-clients unzip \
+  && npm install

--- a/Makefile
+++ b/Makefile
@@ -5,6 +5,8 @@ all: setup build
 
 build:
 	make docker-run CMD="sh ./scripts/pibuilder.sh"
+
+	@echo "Now write ./cache/os.img to an SD card"
 .PHONY: build
 
 docker-build:
@@ -30,4 +32,6 @@ docker-run:
 
 setup:
 	make docker-run CMD="node ./scripts/setup.js" RUN_USER=1000
+
+	@echo "Now run 'make build' to configure the image"
 .PHONY: setup

--- a/Makefile
+++ b/Makefile
@@ -1,9 +1,33 @@
+DOCKER_CONTAINER ?= riggerthegeek/pibuilder
+RUN_USER ?= 0
+
 all: setup build
 
 build:
-	sh ./scripts/pibuilder.sh
+	make docker-run CMD="sh ./scripts/pibuilder.sh"
 .PHONY: build
 
+docker-build:
+	docker build -t ${DOCKER_CONTAINER} .
+.PHONY: docker-build
+
+docker-run:
+	touch settings.sh
+	mkdir -p ./cache
+	mkdir -p ./ssh-keys
+	docker run \
+		-it \
+		--privileged \
+		--rm \
+		-v "${PWD}/settings.sh:/opt/pibuilder/settings.sh" \
+		-v "${PWD}/cache:/opt/pibuilder/cache" \
+		-v "${PWD}/ssh-keys:/ssh-keys" \
+		-v "${PWD}/scripts:/opt/pibuilder/scripts" \
+		-u ${RUN_USER} \
+		${DOCKER_CONTAINER} \
+		${CMD}
+.PHONY: docker-run
+
 setup:
-	node ./scripts/setup.js
+	make docker-run CMD="node ./scripts/setup.js" RUN_USER=1000
 .PHONY: setup

--- a/README.md
+++ b/README.md
@@ -18,7 +18,7 @@ A script to automate the headless building of a configured Raspberry Pi
 
 # Config
 
-> Requires NodeJS v4 or above
+> Requires Docker v1.17 or above
 
 The recommended way is to use the NodeJS script to build the `./settings.sh`
 file for you. This will ask you various questions about how you want your
@@ -67,8 +67,7 @@ the **PI_USERNAME** as part of the `docker` group. The command it runs is:
 # Run
 
 Once you have got your settings, hit `sudo make build`. This will create
-a new Pi image you can flash to a memory card (suggest
-[Etcher](http://etcher.io)).
+a new Pi image you can flash to a memory card (suggest [Etcher](http://etcher.io)).
 
 When you power-up the Pi, which will take between 30 seconds and 5 minutes 
 (the longest part is installing Docker), you will eventually be able to connect 
@@ -83,8 +82,8 @@ will send you an email when it's all configured.
 # ToDo
 
 - [ ] Get a web app set up to allow `.img` file to be done remotely
-- [ ] Get working on other operating systems (currently only tested on Fedora)
 - [ ] Watch for changes in a Git repo to allow automated deployments to the Pi 
+- [x] Get working on other operating systems (currently only tested on Fedora)
 - [x] Cronjob to do updates
 - [x] Get `make setup` to prompt for the setup
 - [x] Specify an IP address range

--- a/scripts/pibuilder.sh
+++ b/scripts/pibuilder.sh
@@ -128,8 +128,6 @@ sed -i "s/%PI_IP_ADDRESS_RANGE_START%/$PI_IP_ADDRESS_RANGE_START/g" "${ROOT_DIR}
 sed -i "s/%PI_IP_ADDRESS_RANGE_END%/$PI_IP_ADDRESS_RANGE_END/g" "${ROOT_DIR}/first_run.sh"
 sed -i "s/%PI_DNS_ADDRESS%/$PI_DNS_ADDRESS/g" "${ROOT_DIR}/first_run.sh"
 
-cat "${ROOT_DIR}/first_run.sh"
-
 chmod 755 "${ROOT_DIR}/first_run.sh"
 cp "${PI_SSH_KEY}" "${ROOT_DIR}/id_rsa.pub"
 cp -Rf ./data "${ROOT_DIR}/opt/data"

--- a/scripts/setup.js
+++ b/scripts/setup.js
@@ -13,13 +13,16 @@ const fs = require('fs.extra');
 const inquirer = require('inquirer');
 const inquirerFilePath = require('inquirer-file-path');
 const ipToInt = require('ip-to-int');
-const keyGen = require('ssh-keygen')
+const keyGen = require('ssh-keygen');
 const validateEmail = require('email-validator');
 const validateIp = require('validate-ip');
 
 /* Files */
 
 inquirer.registerPrompt('filePath', inquirerFilePath);
+
+const keyDir = process.env.SSH_KEY_DIR || path.join(__dirname, '..', 'ssh-keys');
+fs.mkdirpSync(keyDir);
 
 const Validate = {
   integer: input => {
@@ -174,7 +177,7 @@ const questions = [{
   when: answers => answers._useWifi,
   validate: Validate.required,
 }, {
-  type: 'password',
+  type: 'input',
   name: 'PI_WIFI_PASS',
   message: 'WiFi Password',
   when: answers => answers._useWifi,
@@ -280,9 +283,6 @@ inquirer.prompt(questions)
       return answers;
     }
 
-    const keyDir = path.join(__dirname, '..', 'ssh-keys');
-    fs.mkdirpSync(keyDir);
-
     const location = path.join(keyDir, answers.PI_HOSTNAME);
 
     return new Promise((resolve, reject) => {
@@ -322,4 +322,11 @@ inquirer.prompt(questions)
     const data = settings.join('\n');
 
     fs.writeFileSync(filePath, data, 'utf8');
+
+    console.log('Setup completed successfully');
+  })
+  .catch((err) => {
+    console.log('Setup errored');
+    console.log(err.stack || err);
+    process.exit(1);
   });

